libmscore/changeMap.cpp

```diff
@@ -0,0 +1,487 @@
+            case ChangeMethod::NORMAL:
+            default:
+                  valueFunction = [&](int ct) { return int(
+                        double(exprDiff) * (double(ct) / double(exprTicks))
+                        ); };
+                  break;
+            }
+
+      return event.cachedStart + valueFunction(tick.ticks() - event.tick.ticks());
+      }
+
+//---------------------------------------------------------
+//   val
+///   return value at tick position. Do not confuse with
+///   `value`, which is a method of QMultiMap.
+//---------------------------------------------------------
+
+int ChangeMap::val(Fraction tick)
+      {
+      if (!cleanedUp)
+            cleanup();
+
+      auto eventIter = upperBound(tick);
+      if (eventIter == begin()) {
+            return DEFAULT_VALUE;
+            }
+
+      eventIter--;
+      bool foundRamp = false;
+      ChangeEvent& rampFound = eventIter.value();       // only used to init
+      for (auto& event : values(eventIter.key())) {
+            if (event.type == ChangeEventType::RAMP) {
+                  foundRamp = true;
+                  rampFound = event;
+                  }
+            }
+
+      if (!foundRamp) {
+            // Last event must be a fix, since there are max two events at one tick
+            return eventIter.value().value;
+            }
+
+      if (tick >= rampFound.etick) {
+            return rampFound.cachedEnd;
+            }
+      else {
+            // Do some maths!
+            return interpolate(eventIter.value(), tick);
+            }
+      }
+
+//---------------------------------------------------------
+//   addFixed
+//---------------------------------------------------------
+
+void ChangeMap::addFixed(Fraction tick, int value)
+      {
+      insert(tick, ChangeEvent(tick, value));
+      cleanedUp = false;
+      }
+
+//---------------------------------------------------------
+//   addRamp
+///   A `change` of 0 means that the change in velocity should be calculated from the next fixed
+///   velocity event.
+//---------------------------------------------------------
+
+void ChangeMap::addRamp(Fraction stick, Fraction etick, int change, ChangeMethod method, ChangeDirection direction)
+      {
+      change = abs(change);
+      change *= (direction == ChangeDirection::INCREASING) ? 1 : -1;
+      insert(stick, ChangeEvent(stick, etick, change, method, direction));
+      cleanedUp = false;
+      }
+
+//---------------------------------------------------------
+//   cleanup
+//---------------------------------------------------------
+
+void ChangeMap::cleanup()
+      {
+      //
+      // Stage 0: put the ramps in size order if they start at the same point
+      //
+      for (auto& tick : uniqueKeys()) {
+            // `events` will contain all the ramps at this tick
+            std::vector<ChangeEvent> events;
+            for (auto& event : values(tick)) {
+                  if (event.type == ChangeEventType::FIX)
```

Comment:
Also instead of skipping `FIX` events we could just call push_back for `RAMP` events. This is a minor suggestion, of course, but could still make this code more explicit about describing what it is intended for.